src/mem/ruby/structures/ALUFreeListArray.hh

```diff
@@ -0,0 +1,51 @@
+#ifndef __MEM_RUBY_STRUCTURES_ALUFREELISTARRAY_HH__
+#define __MEM_RUBY_STRUCTURES_ALUFREELISTARRAY_HH__
+
+#include <deque>
+
+#include "mem/ruby/common/TypeDefines.hh"
+#include "mem/ruby/system/RubySystem.hh"
+
+namespace gem5
+{
+
+namespace ruby
+{
+
+class ALUFreeListArray
+{
+  private:
+    unsigned int alus;
+    Cycles accessLatency;
+    RubySystem *m_ruby_system;
+
+    class AccessRecord
+    {
+      public:
+        AccessRecord() : idx(0), startTick(0) {}
+        AccessRecord(int64_t idx, Tick startTick) {
+          this->idx = idx;
+          this->startTick = startTick;
+        }
+
+        int64_t idx;
+        Tick startTick;
+    };
+
+    // Queue of accesses from past accessLatency cycles
+    std::deque<AccessRecord> accessQueue;
+
+  public:
+    ALUFreeListArray(unsigned int alus, Cycles accessLatency, RubySystem *rs);
+
+    bool tryAccess(int64_t idx);
```

Comment:
What is `idx` in this case? Are the ALUs used for atomic tagged based on the set index?

I think that you don't want to have any parameter here and that you want the ALUs to be, in some sense, fully associative. I could be wrong here, though.

Alternatively, if you want to model that there is 1 ALU per cache bank, you should probably pass the full address in and have the notion of bank mapping in this object.

Reply:
@DanKouch can chime in, but ultimately I believe the idx is representing a bank of the L2 (or a pseudo-bank).  I thought some of the other code in this file was doing the bank mapping -- Dan can correct me if I'm wrong here.  Yes we want the ALUs to be fully associative, but because we need to ensure a given cache line waits the full time while the ALU is allowed to be reused for other cache lines to mimic the pipeline behavior of real GPUs.  So there will need to be some parameter, but we can certainly debate what the parameter is.

Reply:
Right now, the `idx` that is passed into this function is the set number. I don't think that is the intended behavior. It should probably be the bank number, if that's the intention. The other option, which may be easier, would be to just pass in the address and do all of the bank and associativity logic in this object.

Reply:
The intent here is to keep track of what cache lines are currently in an atomic ALU pipeline. Since the ALUs are intended to be fully-assosciative, we don't need the bank number. In the latest push I've updated the class to take the full address, then determine the cache line from there.

Reply:
See comment on updated patch